### PR TITLE
Breaking: Spawn webserver in separate process

### LINKED ISSUE #1694: [Tests] Run test web server(s) in a separate thread
# 🚀 Feature request

## Description

Right now the web server created by `hintRunner` runs in the same process.
Spawn the web server into its own process. Right now the web server and the tests run in the same process and I think that some of the timeouts we are getting are because of that.

**Are you willing to work on this?** 😏

## What scenarios will this solve?

This should hopefully make the tests more reliable. I have a feeling that because everything is running in the same thread sometimes we get timeouts when we shouldn't or that we are serving old content.
This will probably have a perf impact because we will be spawning quite a few new processes so we need to be careful how we do this. An option could be to use one process per testing file instead of per test.

### PR DESCRIPTION
**Pull request by molant**
## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [x] Added/Updated related documentation.
- [x] Added/Updated related tests.

## Short description of the change(s)

This currently spawns a new server per test. I was thinking about doing something smart and have just one per test file but ava spawns every test in parallel so I'm not sure that would change anything.

Also this PR is missing:

* [x] Update the testing documentation on how the test server works
* [x] Update the `create-hint` package

But it will be good if you can start taking a look into it. 

Fix #1694

### CODE REVIEW

**Comment by molant** [comment]
BTW, I passed all the tests on Windows for the first time in a very long time 😊

File: package.json [hunk 1]
```diff
@@ -54,6 +54,7 @@
     "pre-release": "npm run clean:packages && npm run build:scripts && npm run release -- --prerelease",
     "release": "npm run clean:packages && npm run build:scripts && node dist/scripts/release.js",
     "test": "npm run clean && node scripts/test.js",
+    "test-hints": "loop \"yarn test-only\" --cwd packages --exclude configuration-development,configuration-progressive-web-apps,configuration-web-recommended, connector-chrome, connector-edge, connector-jsdom, connector-local, create-hint, create-hintrc, create-parser, extension-browser, extension-vscode, formatter-codeframe, formatter-excel, formatter-html, formatter-json, formatter-stylish, formatter-summary, hint, parser-babel-config, parser-css, parser-html, parser-javascript, parser-manifest, parser-package-json, parser-typescript-config, parser-webpack-config, utils-connector-tools, utils-create-server, utils-debugging-protocol-common, utils-tests-helpers",
```
Comment (molant) [inline package.json]: I'll remove this.

File: packages/utils-tests-helpers/src/hint-runner.ts [hunk 2]
```diff
@@ -124,31 +124,26 @@ const validateResults = (t: ExecutionContext<HintRunnerContext>, results: Proble
...
-            return;
-        }
-        t.context.server = createServer(configs.https);
-        await t.context.server.start();
-    });
-
-    test.afterEach.always(async (t) => {
-        await t.context.server.stop();
-    });
+    // /**
```
Comment (molant) [inline packages/utils-tests-helpers/src/hint-runner.ts]: Will delete this.

**Comment by molant** [comment]
@sarvaje I've done your feedback

**Comment by molant** [comment]
* [x] Update the `create-hint` package ➡ Doesn't need any manual update. We mark this as a breaking change so the release script updates everything.

@alrra, did you change the script so major version bumps of `devDependencies` are patches/minors?

**Comment by sarvaje** [comment]
@molant why are you ignoring `connector-chrome` in so many tests?

File: packages/utils-create-server/README.md [hunk 3]
```diff
@@ -9,3 +9,50 @@ To install the package, you need to run:
...
 ```
+
+## Architecture
+
+The test web server is `spawn`ed in a new thread. There reasons for this are:
+
+* Better isolation
+* Improve the debugging experience
+
+With the current architecture lnowing what is actually returned by the
```
Comment (sarvaje) [inline packages/utils-create-server/README.md]:
```suggestion
With the current architecture knowing what is actually returned by the
```

**Comment by molant** [comment]
> why are you ignoring `connector-chrome` in so many tests?

Windows build keep failing on Windows and I'm testing things to make it work. I was expecting them to be more reliable by spawning less chrome instances :(

**Comment by sarvaje** [comment]
> Windows build keep failing on Windows and I'm testing things to make it work. I was expecting them to be more reliable by spawning less chrome instances :(

But it is just for testing, or is it going to be permanent?

**Comment by molant** [comment]
> But it is just for testing, or is it going to be permanent?

🤷‍♂️, I'm doing tests for now

**Comment by alrra** [comment]
>     * [x]  Update the `create-hint` package ➡ Doesn't need any manual update. We mark this as a breaking change so the release script updates everything.
> 
> 
> @alrra, did you change the script so major version bumps of `devDependencies` are patches/minors?

Working on it, will be done later today, or Monday.

For  `create-hint` there shouldn't be any problems, because nothing depends on it.

**Comment by molant** [comment]
I think this is ready for merge. I've done a few changes since the last review so it will be cool if you could review it again:
- Fix a race condition with CDP that was causing the Windows failures (that's what I believe, Windows builds are now passing on Pipelines 🎉)
- Refactor `requestResponse` in jsdom to match the one in chrome (forgot about it last time, both files should look about the same now)
- group some `await`s into `await Promise.all()`

Also this PR should be rebased instead of squashed. I can do it later today if I get enough 👍

File: packages/utils-connector-tools/tests/requester.ts [hunk 4]
```diff
@@ -206,9 +204,9 @@ const loopServerMultiSteps = {
 };
 
 test(`Requester follows all hops, reports the right number and returns the final string content`, async (t) => {
-    const { requester, server } = t.context;
+    const { requester } = t.context;
 
-    server.configure(hopsServerConfig);
+    const server = await Server.create({ configuration: hopsServerConfig });
```
Comment (antross) [inline packages/utils-connector-tools/tests/requester.ts]: Does this need a corresponding `await server.stop()`? I'm only marking this one, but there are multiple occurances in the file.

Comment (molant) [reply packages/utils-connector-tools/tests/requester.ts]: yes, that's why it was failing on Linux and mac at the moment. I've pushed the changes

File: packages/utils-create-server/src/server.ts [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+BBQeQAc/3f6lP2Y6GqG/7KCuteMQ4zAfBgNVHSMEGDAWgBQeQAc/3f6lP2Y6GqG/
+7KCuteMQ4zAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBBQUAA4IBAQBPXdpOHyTn
+UxEXTxBoVSebpewYM/4lyXNI+13j64b2t6wC+nHXZ+beZUMGwznYGe1hXZ0fxfS+
+NDEuOEuGvpLReAFj2kW6PpTXg++jt/monBzHZA7RoZxYUJPJImpO/ByEtYoMg5fk
+e3111eFlHH+osZ9I1kNmJAAvzNHlnTBIk0sxBRz9jW7hkLJkC0s1pREAwr/sKPjy
+xMV8++vDR/e5RfJDHmqCmSv7DG5Q1cLBjvb09Kn9mSqAi9iuZvbMx+mxkocEFC31
+w4C7/uFcTns/5rYjv+vp+EGv5cupxJjXpd1O/buT/Tt/Ur5maySbu4vDYVgmPuXU
+7avEZJTeYgih
+-----END CERTIFICATE-----`
+    private static readonly _key: string =
```
[the remark this answers is not in the document]
Comment (molant) [reply packages/utils-create-server/src/server.ts]: Perf, that way we avoid reading from disk and it's not like we are going to change them.

**Comment by molant** [comment]
> Looks good to me with the updates assuming tests pass.

Tests pass! mac is failing because of #1772 which happens randomly but that's unrelated to this PR.

File: packages/utils-create-server/src/server.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/utils-create-server/src/server.ts]: Ok!

**Pull request merged**